src/test/java/educonnect/logic/commands/CopyCommandTest.java

```diff
@@ -0,0 +1,98 @@
+    public void equals() {
+        TagContainsKeywordsPredicate firstPredicate =
+                new TagContainsKeywordsPredicate(new Tag("first"));
+        TagContainsKeywordsPredicate secondPredicate =
+                new TagContainsKeywordsPredicate(new Tag("second"));
+
+        CopyCommand copyFirstCommand = new CopyCommand(List.of(firstPredicate));
+        CopyCommand copySecondCommand = new CopyCommand(List.of(secondPredicate));
+
+        // same object -> returns true
+        assertTrue(copyFirstCommand.equals(copyFirstCommand));
+
+        // same values -> returns true
+        CopyCommand findFirstCommandCopy = new CopyCommand(List.of(firstPredicate));
+        assertTrue(copyFirstCommand.equals(findFirstCommandCopy));
+
+        // different types -> returns false
+        assertFalse(copyFirstCommand.equals(1));
+
+        // null -> returns false
+        assertFalse(copyFirstCommand.equals(null));
+
+        // different student -> returns false
+        assertFalse(copyFirstCommand.equals(copySecondCommand));
+    }
+
+    @Test
+    public void execute_zeroKeywords_allEmailsCopied() {
+        expectedModel.updateFilteredStudentList(Model.PREDICATE_SHOW_ALL_STUDENTS);
+        ObservableList<Student> students = expectedModel.getFilteredStudentList();
+
+        String expectedMessage = String.format(Messages.MESSAGE_STUDENT_EMAIL_COPIED_OVERVIEW,
+                students.size());
+        CopyCommand command = new CopyCommand(Collections.emptyList(), false);
+        CommandTestUtil.assertCommandSuccess(command, model, expectedMessage.toString(), expectedModel);
+    }
+
+    @Test
+    public void execute_oneTag_multipleEmailsCopied() {
+        TagContainsKeywordsPredicate predicate = preparePredicate("tutorial-2");
+        expectedModel.updateFilteredStudentList(List.of(predicate));
+        ObservableList<Student> students = expectedModel.getFilteredStudentList();
+
+        String expectedMessage = String.format(Messages.MESSAGE_STUDENT_EMAIL_COPIED_OVERVIEW, students.size());
+        CopyCommand command = new CopyCommand(List.of(predicate), false);
+        CommandTestUtil.assertCommandSuccess(command, model, expectedMessage, expectedModel);
+        assertEquals(Arrays.asList(TypicalStudents.BENSON,
+                TypicalStudents.FIONA), model.getFilteredStudentList());
+    }
+
+    @Test
+    public void toStringMethod() {
+        List<Predicate<Student>> predicates = List.of(new NameContainsKeywordsPredicate("keyword"));
+        CopyCommand copyCommand = new CopyCommand(predicates);
+        String expected = CopyCommand.class.getCanonicalName() + "{predicates=" + predicates + "}";
+        assertEquals(expected, copyCommand.toString());
+    }
+
+    /**
+     * Parses {@code userInput} into a {@code TagContainsKeywordsPredicate}.
+     */
+    private TagContainsKeywordsPredicate preparePredicate(String userInput) {
+        return new TagContainsKeywordsPredicate(new Tag(userInput));
+    }
+}
```

Comment:
Good to see many test cases here

Reply:
I tried to cover as many cases as I could think of 😃 

For the `CopyCommandTest`, I did not manage to test the clipboard function as tests run on a separate `test thread` instead of the `main thread`. This causes the JavaFX Clipboard library to throw a `InvalidStateException`. From some searching, I do not think testing the library functions are required anyway (they should be working as intended).